### ISSUE TITLE
Missed packets and no data on cortex-a53

### ISSUE THREAD

**Issue by dkargin** [xqms/xtion#17]
Good day
I am trying to run your driver on odroid-c2 sbc. 

When I run xtion-grabber, I get following spam in syslog:

```
[ 1322.775707] usb 1-1.4: Missed packets: 2750 -> 2752
[ 1322.780566] usb 1-1.4: Missed packets: 2756 -> 2758
[ 1322.785361] usb 1-1.4: Missed packets: 2758 -> 2760
[ 1322.790223] usb 1-1.4: Missed packets: 56618 -> 56620
[ 1322.795198] usb 1-1.4: Missed packets: 2766 -> 2768
[ 1322.800060] usb 1-1.4: Missed packets: 2772 -> 2774
[ 1322.804898] usb 1-1.4: Missed packets: 56638 -> 56640
[ 1322.809898] usb 1-1.4: Missed packets: 2777 -> 2779
[ 1322.814689] usb 1-1.4: Missed packets: 2779 -> 2781
[ 1322.819574] usb 1-1.4: Missed packets: 2787 -> 2789
[ 1322.824360] usb 1-1.4: Missed packets: 2795 -> 2797
[ 1322.829228] usb 1-1.4: Missed packets: 2800 -> 2802
[ 1322.834012] usb 1-1.4: Missed packets: 2802 -> 2805
[ 1322.838890] usb 1-1.4: Missed packets: 56670 -> 56672
[ 1322.843868] usb 1-1.4: Missed packets: 2809 -> 2811
[ 1322.848679] usb 1-1.4: Missed packets: 2811 -> 2813
[ 1322.853509] usb 1-1.4: Missed packets: 2819 -> 2821
[ 1322.858374] usb 1-1.4: Missed packets: 2824 -> 2826
[ 1322.863162] usb 1-1.4: Missed packets: 2826 -> 2829
[ 1322.868041] usb 1-1.4: Missed packets: 2833 -> 2835
[ 1322.872822] usb 1-1.4: Missed packets: 2835 -> 2837
[ 1322.877691] usb 1-1.4: Missed packets: 2841 -> 2843
[ 1322.882479] usb 1-1.4: Missed packets: 2843 -> 2845
[ 1322.887345] usb 1-1.4: Missed packets: 56702 -> 39526
[ 1322.892349] usb 1-1.4: Missed packets: 2848 -> 2850
[ 1322.897146] usb 1-1.4: Missed packets: 2850 -> 2852
[ 1322.902013] usb 1-1.4: Missed packets: 2856 -> 2858
[ 1322.906804] usb 1-1.4: Missed packets: 2858 -> 2860
[ 1322.911637] usb 1-1.4: Missed packets: 2866 -> 2868
[ 1322.916504] usb 1-1.4: Missed packets: 2871 -> 2873
```

No frame data is obtained.

It works fine on x86/x64 with the same device. There are still some missing packets (with completely random ID->ID), but xtion-grabber still works. 

Kernel version:

```
Linux odroid-c2 3.14.29-odroid #1 SMP PREEMPT Thu Jul 21 12:40:43 EDT 2016 aarch64 aarch64 aarch64 GNU/Linux
```

I use [fork](https://github.com/FrostHand/xtion/tree/yocto) that is compatible with 3.14.x kernels and yocto (no vb2_v4l_buffer) but I guess this problem is quite generic.

**Comment by xqms** [comment]
Hi,

did you try another driver (i.e. OpenNI2) on this device?
Usually, this problem occurs under high CPU load. Maybe your USB host cannot cope with the data rate from the Xtion? Since only few packets are lost in between the received packets, this seems to be the case here.

You can try playing with `XTION_NUM_URBS` and `XTION_URB_SIZE` in xtion.h. It seems that there are different optimum values for different USB host controllers, so you might have luck with different settings.

**Comment by dkargin** [comment]
Yep, I've checked OpenNI2 already, it works.

The same problem was with Cortex-A15 board (Odroid-XU4). There were no problems with CPU and OpenNI2-based grabber, but your kernel driver was loosing packets and giving no data.

Right now I have the following values in xtion.h :

```
#define XTION_NUM_URBS 16
// #define XTION_URB_SIZE (6 * 20480UL)
// #define XTION_URB_SIZE 81920
#define XTION_URB_SIZE 20480
```

Should I increase this values?

**Comment by xqms** [comment]
Yes, you can try higher values both for the URB size and for the number of URBs.

If you can provide it, I would be interested in some USB sniffer data from an OpenNI run on your hardware. You can capture such data with wireshark (after loading the `usbmon` kernel driver).

**Comment by dkargin** [comment]
Yep, it works. There are still missed packets, but spam is over.

```
#define XTION_URB_SIZE (6 * 20480UL)
#define XTION_NUM_URBS 32
```

But XtionGrabber::read_thread() now stops on 

```
if(ioctl(m_color_fd, VIDIOC_DQBUF, &buf) != 0)
{
   ...
   return;
}
```

The most interesting that errno=0 ("No error") in that case

**Comment by xqms** [comment]
Hm, that should not happen. Do you see anything interesting in dmesg?
I would still be interested in a USB capture from a successful OpenNI run to see what URB sizes etc. they use.

**Comment by dkargin** [comment]
Right, color grabbing barely works, but depth is OK. I've reordered grabbing thread a bit (in the xtion_grabber). It does not exit on first error. 

Checking data with rostopic hz -w 10 ... . It gets about 25-29Hz for /camera/depth*, but when I try to get some color, it gets about 2-5Hz, and a lot of errors like 

```
Could not queue color buffer, errno=22: Invalid argument
```

It goes from lines:

```
if(ioctl(m_color_fd, VIDIOC_QBUF, &buffer->buf) != 0)
{
    NODELET_INFO("Could not queue color buffer, errno=%d: %s", errno, strerror(errno));
    return -1;
}
```

And It seems all external communication (usb bus, OTG bus with g_multi driver, and ethernet as well) goes down when I try to get color data. It restores back when I unplug xtion.

Do you have exact command for usbmon?

**Comment by dkargin** [comment]
I'll need some time to get usbmon into our linux distribution (yocto)

**Comment by dkargin** [comment]
[xtion.dump.zip](https://github.com/xqms/xtion/files/588711/xtion.dump.zip)
Finally I've prepered usbmon dump for xtion, while it was working with openni2-based grabber
